cli/cmd/migrate.go

```diff
@@ -0,0 +1,137 @@
+cloudquery migrate ./directory ./aws.yml ./pg.yml
+`
+)
+
+func NewCmdMigrate() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "migrate [files or directories]",
+		Short:   migrateShort,
+		Long:    migrateShort,
+		Example: migrateExample,
+		Args:    cobra.MinimumNArgs(1),
+		RunE:    migrate,
+	}
+	return cmd
+}
+
+func migrate(cmd *cobra.Command, args []string) error {
+	cqDir, err := cmd.Flags().GetString("cq-dir")
+	if err != nil {
+		return err
+	}
+
+	ctx := cmd.Context()
+	log.Info().Strs("args", args).Msg("Loading spec(s)")
+	fmt.Printf("Loading spec(s) from %s\n", strings.Join(args, ", "))
+	specReader, err := specs.NewSpecReader(args)
+	if err != nil {
+		return fmt.Errorf("failed to load spec(s) from %s. Error: %w", strings.Join(args, ", "), err)
+	}
+
+	for _, sourceSpec := range specReader.Sources {
+		if len(sourceSpec.Destinations) == 0 {
+			return fmt.Errorf("no destinations found for source %s", sourceSpec.Name)
+		}
+		var destinationsSpecs []specs.Destination
+		for _, destination := range sourceSpec.Destinations {
+			spec := specReader.Destinations[destination]
+			if spec == nil {
+				return fmt.Errorf("failed to find destination %s in source %s", destination, sourceSpec.Name)
+			}
+			destinationsSpecs = append(destinationsSpecs, *spec)
+		}
+		if err := migrateConnection(ctx, cqDir, *sourceSpec, destinationsSpecs); err != nil {
+			return fmt.Errorf("failed to migrate source %s: %w", sourceSpec.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func migrateConnection(ctx context.Context, cqDir string, sourceSpec specs.Source, destinationsSpecs []specs.Destination) error {
+	destinationNames := make([]string, len(destinationsSpecs))
+	for i := range destinationsSpecs {
+		destinationNames[i] = destinationsSpecs[i].Name
+	}
+	migrateTime := time.Now().UTC()
+
+	log.Info().Str("source", sourceSpec.Name).Strs("destinations", destinationNames).Time("migrate_time", migrateTime).Msg("Start migration")
+	defer log.Info().Str("source", sourceSpec.Name).Strs("destinations", destinationNames).Time("migrate_time", migrateTime).Msg("End migration")
+
+	sourceClient, err := clients.NewSourceClient(ctx, sourceSpec.Registry, sourceSpec.Path, sourceSpec.Version,
+		clients.WithSourceLogger(log.Logger),
+		clients.WithSourceDirectory(cqDir),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get source plugin client for %s: %w", sourceSpec.Name, err)
+	}
+	defer func() {
+		if err := sourceClient.Terminate(); err != nil {
+			log.Error().Err(err).Msg("Failed to terminate source client")
+			fmt.Println("failed to terminate source client: ", err)
+		}
+	}()
+
+	destClients, err := initializeDestinationClients(ctx, sourceSpec, destinationsSpecs, cqDir)
+	// we make sure to defer the closing of clients here before checking the error, otherwise
+	// if one of the clients returns an error, the clients opened before it may never get closed
+	defer func() {
+		for _, destClient := range destClients {
+			if destClient != nil {
```

Comment:
Wouldn't really work because if you look at `initializeDestinationClients`, partly-initialized slice isn't returned if there was an error.

Reply:
Ah, nice catch. My plan was to return the slice from that function, but looks like I forgot to do that! 🤦

Reply:
Updated, PTAL 🙇